changefeedccl: protect lagging tables with its own timestamp

aerfrei opened the pull request:
Before this change, changefeeds watching multiple tables
would protect all watched tables and system tables with a
single protected timestamp (pts) record. All tables would
be protected back to the highwater for the changefeed.

In the event where one table lags behind the others, this
would result in preventing more garbage collection than
needed for many tables. This would become more of an issue
when we introduce DB-level feeds because they will be
implemented with multi-table feeds.

We now protect lagging tables separately. We use per table
highwaters to identify tables who are lagging behing the
most advanced table by more than the new
changefeed.protect_timestamp_bucketing_interval cluster
setting. All tables at least that far behind will be protected
by their own pts record.

Epic: CRDB-1421
Fixes: #147785

Release note (performance): Multi-table changefeeds will protect
each lagging user table from garbage collection with its own record.
This will improve performance of multi-table changefeeds including
upcoming db-level changefeeds by allowing more garbage collection.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149616)
<!-- Reviewable:end -->

aerfrei committed:
changefeedccl: protect lagging tables with its own timestamp

Before this change, changefeeds watching multiple tables
would protect all watched tables and system tables with a
single protected timestamp (pts) record. All tables would
be protected back to the highwater for the changefeed.

In the event where one table lags behind the others, this
would result in preventing more garbage collection than
needed for many tables. This would become more of an issue
when we introduce DB-level feeds because they will be
implemented with multi-table feeds.

We now protect lagging tables separately. We use per table
highwaters to identify tables who are lagging behing the
most advanced table by more than the new
changefeed.protect_timestamp_bucketing_interval cluster
setting. All tables at least that far behind will be protected
by their own pts record.

Epic: CRDB-1421
Fixes: #147785

Release note (performance): Multi-table changefeeds will protect
each lagging user table from garbage collection with its own record.
This will improve performance of multi-table changefeeds including
upcoming db-level changefeeds by allowing more garbage collection.

aerfrei commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -1225,6 +1225,17 @@ message ChangefeedProgress {
+  //
+  // The same cleanup and reconciliation mechanisms apply as with
+  // ProtectedTimestampRecord.
+  repeated bytes protected_timestamp_records = 6 [
```
Yes, that should be in the second commit. Hoping to squash these soon. Let me know what you think of it. I'm using the WriteChunkedFileToJobInfo which will write this in potentially multiple rows to the job_info table

andyyang890 commented on pkg/ccl/changefeedccl/changefeedbase/settings.go:
```diff
@@ -215,6 +215,16 @@ var ProtectTimestampLag = settings.RegisterDurationSetting(
+	"if true, creates separate protected timestamp records for each table in a changefeed; "+
+		"if false, uses a single protected timestamp record for all tables",
+	false,
+	settings.WithPublic)
```
I think we generally don't document settings like this but probably worth asking the team in Slack

andyyang890 reviewed (commented):
Have a lot of the same comments/questions as Miles

aerfrei commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
Ok, I'll check that out. What does this mean for multi-table feeds that aren't DB or Schema level? Should I implement this separately for both?

aerfrei commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
I think I was confused by the existing behavior immediately below which doesn't seem to check for existing PTS records. I think in my testing I was only thinking about the case where this is running for a newly created changefeed. Does the changefeedPlanHook also run when we resume paused changefeeds?

aerfrei commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -11197,6 +11197,132 @@ func TestHighwaterDoesNotRegressOnRetry(t *testing.T) {
+// TestChangefeedPerTableProtectedTimestampAdvancement tests that when per-table
+// protected timestamps are enabled, all individual PTS records advance to exactly
+// the same timestamp as the changefeed's high watermark.
+func TestChangefeedPerTableProtectedTimestampAdvancement(t *testing.T) {
```
Yes, good point, hoping not to merge that until after that work is in. I'm hoping at that point this test setup will remain the same but the assertions will be different.

aerfrei commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -1733,3 +1733,29 @@ message RetriableExecutionFailure {
+// The same cleanup and reconciliation mechanisms apply as with
+// ProtectedTimestampRecord.
+message ProtectedTimestampRecords {
+  repeated PerTableProtectedTimestampRecords per_table_protected_timestamp_records = 1 [(gogoproto.nullable) = false];
```
Yes, I think that's a good idea. Working on that and I'll include a name change with that

aerfrei commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1895,6 +1909,27 @@ func (cf *changeFrontier) manageProtectedTimestamps(
 		highWater = cf.highWaterAtStart
 	}
 
+	if len(ptsEntries.ProtectedTimestampRecords) > 0 {
```
@andyyang890 Right now, this code is using the pts entries as a kind of proxy to wether the changefeed is using the new PTS system. Should this instead be checking a ChangefeedProgressConfig field?

aerfrei replied:
This is a good point! There's another jira issue to deal more with conversion, so in the interests of keeping this PR small and getting it merged, I'm getting rid of this logic and I'll call a function from cf.Start() when I implement the switching between the formats in a later PR.

andyyang890 commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -1733,3 +1733,11 @@ message RetriableExecutionFailure {
 }
+
+// ProtectedTimestampRecords is a map from table descriptor IDs to protected timestamp record IDs.
+message ProtectedTimestampRecords {
```
@aerfrei We discussed this offline already, but just noting here for posterity that `pkg/ccl/changefeedccl/changefeedpb/changefeed.proto` would be a better location for this protobuf.

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1895,6 +1904,31 @@ func (cf *changeFrontier) manageProtectedTimestamps(
+			return false, cf.knobs.ManagePTSError()
+		}
+		for _, record := range ptsEntries.PerTableProtectedTimestampRecords {
+			recId := record.ProtectedTimestampRecordID
```
Ah sorry for not being clear, I meant replace `recId` with `recID`

andyyang890 commented on pkg/ccl/changefeedccl/changefeedbase/settings.go:
```diff
@@ -215,6 +215,16 @@ var ProtectTimestampLag = settings.RegisterDurationSetting(
+var PerTableProtectedTimestamps = settings.RegisterBoolSetting(
+	settings.ApplicationLevel,
+	"changefeed.protected_timestamp.per_table.enabled",
+	"if true, creates separate protected timestamp records for each table in a changefeed; "+
```
Don't forget to also update the release note in the PR desc any time you update it in the commit message! 😄

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -12047,6 +12049,138 @@ func TestChangefeedProtectedTimestampUpdate(t *testing.T) {
+// protected timestamp records when PerTableProtectedTimestamps is disabled,
+// that it will NOT create per-table protected timestamp records, and that
+// it will increment the relevant metrics when managing its protected timestamps.
+func TestChangefeedProtectedTimestampUpdateForMultipleTables(t *testing.T) {
```
Seems like this is failing under stress on CI

aerfrei replied:
Sorry about that, I had thought I fixed it but working more on that now. Should be resolved on my next push.

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_job_info.go:
```diff
@@ -25,6 +25,8 @@ const (
 const (
 	// resolvedTablesFilename: ~changefeed/resolved_tables.binpb
 	resolvedTablesFilename = jobInfoFilenamePrefix + "resolved_tables" + jobInfoFilenameExtension
+	// perTableProtectedTimestampsFilename: ~changefeed/per_table_protected_timestamps.binpb
```
I realize that this isn't true for the resolved tables filename (maybe you can change that too?) but could you replace the underscores with hyphens? Seems like that's the convention in the codebase

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_dist.go:
```diff
@@ -461,6 +461,10 @@ func makePlan(
+				// If the per table pts flag was turned on between changefeed creation and now,
+				// the per table pts records will be rewritten in the new format when the
+				// highwater mark is updated in manageProtectedTimestamps.
+				PerTableProtectedTimestamps: changefeedbase.PerTableProtectedTimestamps.Get(sv),
```
As I mentioned in https://github.com/cockroachdb/cockroach/pull/149616#discussion_r2274260510, we should only consider enabling this if `PerTableTracking` is true

aerfrei commented on pkg/ccl/changefeedccl/changefeed_processors.go:
Yes, it will include every table, but not all the way back to the changefeed's highwater which is the benefit.

Let's say we had n tables that are up to date and their resolved ts is about now, 3pm, but table n+1 is lagging by an hour. We'd end up with two pts records, one that protects all tables to ~3pm (including n+1) and one that protects just n+1 back to 2pm. n+1 is double protected, but we avoided protecting all tables back to 2pm. This is the benefit that per-table PTS gives us and we still have that here. Including table n+1 in that primary record hopefully means less complete rewrites of this pts record and hopefully the code is somewhat simpler.

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1941,16 +1948,214 @@ func (cf *changeFrontier) manageProtectedTimestamps(
+	progress *jobspb.ChangefeedProgress,
+	ptsEntries *changefeedpb.ProtectedTimestampRecords,
+) (newPtsTimestamp hlc.Timestamp, updatedPerTablePts bool, err error) {
+	// TODO: This seems weird, come back to it. Is there a better way to do this?
```
We discussed this offline and one alternative is just to set the "bucket" PTS to be `current time - bucketing interval` so that we never have to worry about regressing it

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
Also discussed offline but leaving a note here that I continue to feel strongly that we should not include all the system tables in every per-table PTS record that we create but rather have one system-table-only PTS record like I commented in the design doc. If you decide not to do that in this PR, could you file an issue for it and leave TODO comments with the issue number in the code?

msbutler replied:
Totally fine to leave it as a seperate todo, so long as no production changefeed will ever enter this immediate state, where there are O(numTable) pts's on the system tables.

aerfrei commented on pkg/ccl/changefeedccl/changefeed_processors.go:
Can I pass nil in as the argument to this write (instead of ptsEntries)? I was about to write this helper in a way that would have been equivalent to that. Do you think it's generally helpful to have a separate function called "deleteChangefeedJobInfo" or something to that effect for readability? My sense is that write nil should be about as clear and is less repetitive.

aerfrei commented on pkg/ccl/changefeedccl/changefeed_processors.go:
Sounds good, instead of having that perTablePTSUpdated bool, I'm not going to potentially write to the job info multiple times, once whenever we create or release an existing record. And that will happen inside release/createPerTableRecord.

aerfrei commented on pkg/ccl/changefeedccl/changefeed_processors.go:
Now that I'm not using the bool above, that would result in n calls to write changefeed jobs (n = number of lagging tables). Which seems fine, I just wanted to call it out. Putting in the change.

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
nit: rename to `advanceProtectedTimestamp`, the `ToTimestamp` part is implicit in the function signature

andyyang890 reviewed (commented):
This PR is starting to look really good!

andyyang890 reviewed (commented):
Prod code looks good, have some comments for the test code

aerfrei commented on pkg/ccl/changefeedccl/protected_timestamps_test.go:
I made this change, I think it does look nicer. Marking as resolved.

andyyang890 reviewed (approved):
Nice work! Thanks for your patience with all the feedback
